pymc3/optimization.py

```diff
@@ -0,0 +1,76 @@
+import tqdm
+import numpy as np
+import theano
+from theano.configparser import change_flags
+import pymc3 as pm
+
+__all__ = [
+    'Optimizer'
+]
+
+
+class Optimizer(object):
+    """
+    Optimization with posterior replacements
+    Parameters
+    ----------
+    approx : Approximation
+    loss : scalar
+    wrt : shared params
+    more_replacements : other replacements in the graph
+    optimizer : callable that returns updates, pm.adam by default
+    """
+    @change_flags(compute_test_value='off')
+    def __init__(self, approx, loss, wrt, more_replacements=None, optimizer=None):
+        if optimizer is None:
+            optimizer = pm.adam
+        self.optimizer = optimizer
+        self.wrt = wrt
+        self.approx = approx
+        self.loss = approx.apply_replacements(loss, more_replacements=more_replacements)
+        updates = self.optimizer(self.loss, self.wrt)
+        self.step_function = theano.function([], self.loss, updates=updates)
+        self.hist = np.asarray(())
+
+    @change_flags(compute_test_value='off')
+    def refresh(self, kwargs=None):
+        """
+        Recompile step_function and reset updates
+
+        Parameters
+        ----------
+        kwargs : kwargs for theano.function
```

Comment:
Hmm this could do with a bit more fleshing out in the docstring I think. Maybe with an example.